rfcs/2024-03-22-20170-trace-data-model.md

```diff
@@ -0,0 +1,196 @@
+was concerned with [validating schemas], but never came attempted to define an event schema.  This leaves this RFC
+to establish that data model.
+
+## Cross cutting concerns
+
+- This is part of the effort to support [sending OpenTelemetry traces], it would also support
+  ingesting OpenTelemetry. As well as to provide a path for a canonical model which traces can be
+  translated to/from ideally without loss of fidelity.
+
+## Scope
+
+### In scope
+
+- Define a Vector Trace event model schema.
+- Support for Links between `TraceEvent`s.
+- Defining a mapping between the Vector trace event model to the Datadog trace event (one direction).
+- Defining a mapping between the Vector trace event model and the OpenTelemetry trace event
+  (bi-directional).
+
+### Out of scope
+
+- Change the `TraceEvent` type internally from a re-typing of `LogEvent`.
+- Define the settings and configuration for sinks and sources using the data model.
+
+## Pain
+
+- Without an internal trace model it makes it more difficult to manage adding `sinks`, `sources`,
+  and `transforms` to Vector allowing for translations functions between the various components.
+
+## Proposal
+
+### User Experience
+
+- As a Vector user I would like to be able to ingest and propagate traces using various protocols.
+  This will allow for the greatest flexibility when managing and migrating services of developed by
+  a set of different sources, including closed and open source software.
+
+### Implementation
+
+For the initial Implementation, the data model will be [v1.1.0][otel-proto-110] of the OpenTelemetry Proto.
+
+Therefore the only necessary mapping will be how the Proto data model maps to the `TraceEvent`.
+
+#### Mapping OpenTelemetry Tracing to Vector Data Model
+
+This section reflects similarly to how the current Log `source` for `opentelemetry` works.
+
+As OpenTelemetry is the primary model the this will only describe the data types. The top level
+`TraceEvent` will represent a top level proto `message ResourceSpans`.
+`Value::Object`.  The remaining proto to mappings can be found:
+
+| Proto Type | Vector Type |
+|------------|-------------|
+| `message AnyValue` | `Value` of the inner |
+| `message ArrayValue` | `Value::Array` with elements in `values` |
+| `message KeyValueList` | `Value::Array` with elements in `values` |
+| Any unspecified `message` types | `Value::Object` |
+| All `enum` types | `Value::Integer` |
+| `string` | `Value::Bytes`  |
+| `fixed64` for datetime values | `Value::Timestamp` |
+| `fixed64`,`fixed32`,`uint32` for generic values  | `Value::Integer` |
+| `bytes` | `Value::Bytes` |
+| `repeated` messages | `Value::Array` of `Value::Object` |
+
+The keys for `message` come the field names for `Value::Object`, and MUST follow a "Snake Case" format.
+As an example, instead of `traceId` and `droppedEventCount`, use `trace_id` and `dropped_event_count`.
```

Comment:
```suggestion
The keys for `message` come the field names for `Value::Object`, and MUST follow a "snake case" format.
As an example, instead of `traceId` and `droppedEventCount`, use `trace_id` and `dropped_event_count`.
```